### ISSUE TITLE
Reachability reduction by separating compilation from reachability

### ISSUE THREAD

**Issue by dmlloyd** [qbicc/qbicc#1607]
Right now, everything that gets compiled is considered reachable, which is probably at least a part of why our output is so massive.

We should explore the following:

* Do not enqueue build time class initializers as "reachable" (i.e. stop using `enqueue` which implies `org.qbicc.facts.core.ExecutableReachabilityFacts#IS_INVOKED` on the initializer)
    * Use a different fact flag like `org.qbicc.plugin.reachability.TypeReachabilityFacts#NEEDS_INITIALIZATION` or something
    * Just run initializers in the interpreter only, and let the interpreter continue to eagerly initialize things as well
    * Remove `org.qbicc.driver.ElementInitializer`
* Change `ReachabilityBlockBuilder` so that it is a standalone visitor which is run only when an element is actually reachable from the root set (this would actually be a fairly small change, since it only runs after `finish` anyway) - this could be done pretty easily by registering another action on `IS_INVOKED`
* Change `ClassInitializingBasicBlockBuilder` so that it only marks the exact class as "needs initialization"; let the interpreter recursively initialize so that this logic is in one place rather than duplicated

**Comment by dmlloyd** [comment]
An initial naive implementation of this results in many errors from missing vtable and itable entries; looks like there's more to it than just this. I might poke at it more tomorrow.

**Comment by dgrove-oss** [comment]
Our output is definitely too massive, but we do recompute reachability "from scratch" in the Analyze phase.  So enqueuing something during ADD because it is executed by a `<clinit>` isn't necessarily the final word.   Last time I turned on the statistics, there were maybe 20% fewer methods reachable in ANALYZE vs. ADD, so it isn't entirely broken.

**Comment by dmlloyd** [comment]
Yeah I agree we're definitely reducing the set significantly between ADD and ANALYZE, but not enough. My initial testing with `grepj` show 856 reachable classes after ANALYZE after the naive changes I mentioned above, down from 3802 from a previous run. As a result, compilation is much faster, so I think this is worth pursuing, as well as #951. Granted it's not fully working yet so it's possible that some things are being missed, and maybe this is not a real gain; maybe rerunning ANALYZE a couple of times would provide a similar result (at least in the final output, if not in compilation speed). More experimentation needed...
